File: src/leetcode/e20/ValidParentheses.go
package e20

// LeetCode 20
func isValid(s string) bool {
	stack := make([]rune, 0)
	for _, c := range s {
		if c == '(' || c == '[' || c == '{' {
			stack = append(stack, c)
		} else if len(stack) == 0 {
			return false
		} else if c == ')' && stack[len(stack)-1] == '(' ||
			c == ']' && stack[len(stack)-1] == '[' ||
			c == '}' && stack[len(stack)-1] == '{' {
			stack = stack[:len(stack)-1]
		} else {
			return false
		}
	}
	return len(stack) == 0
}

File: src/leetcode/e20/ValidParentheses.java
/*
 * Copyright (c) 2023-2025 THGEEK.COM. All rights reserved.
 */

package leetcode.e20;

import java.util.Stack;

/**
 * LeetCode 20
 *
 * @author Tao Hong
 */
public class ValidParentheses {
    public boolean isValid(String s) {
        Stack<Character> parenStack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            if (!parenStack.empty()) {
                if ((parenStack.peek() == '(' && s.charAt(i) == ')') ||
                        (parenStack.peek() == '[' && s.charAt(i) == ']') || (parenStack.peek() == '{' && s.charAt(i) == '}')) {
                    parenStack.pop();
                } else {
                    parenStack.push(s.charAt(i));
                }
            } else {
                parenStack.push(s.charAt(i));
            }
        }
        return (parenStack.empty());
    }
}
